refguide-check with pytest as a runner

ev-br opened the pull request:
SciPy has transitioned to running refguide-check modified doctesting via pytest, and here's a matching proposal for NumPy.

The target invocation would be 
```
$ pip install scipy-doctest    # once
$ pytest --pyargs numpy --doctest-collect=api
```

Currently one needs to manually pytest-ignore several dusty corners so the precise incantation for me locally is
```
$ spin run bash
$ cd ..
$ pytest --pyargs numpy --doctest-modules -v --doctest-collect=api --ignore=/home/br/repos/numpy/build-install/usr/lib/python3.11/site-packages/numpy/distutils --ignore=/home/br/repos/numpy/build-install/usr/lib/python3.11/site-packages/numpy/_core/cversions.py --ignore=/home/br/repos/numpy/build-install/usr/lib/python3.11/site-packages/numpy/_pyinstaller --ignore=/home/br/repos/numpy/build-install/usr/lib/python3.11/site-packages/numpy/random/_examples
```

Of course, all these ignores could/should go into a `spin` command, which is on my TODO list.

For comparison:
- an old version of this, not using pytest: https://github.com/numpy/numpy/pull/21962
- scipy version: https://github.com/scipy/scipy/blob/main/scipy/conftest.py#L266
(Most lot of scipy complexity is filtering out warnings though)

ev-br committed:
TST: run smoke-docs if available

ev-br committed:
DOC: fix several doctest errors

ngoldbaum commented on numpy/conftest.py:
```diff
@@ -136,3 +144,42 @@ def weak_promotion(request):
+        msg = "|".join(msgs)
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings('ignore', category=DeprecationWarning, message=msg)
```
The full tests have a check that doesn't like this, although the error mesage is bad, I think it's supposed to say something like we don't want to let code in that might filter a real warning. Obviously you need to ignore warnings to implement this context manager. Maybe there's a way to dynamically add `pytest.mark.filterwarnings` marks to a test?

ev-br replied:
Let me do the simplest thing and skip the top level conftest.py in that test.

ev-br committed:
TST: skip conftest.py in test_warnings

ev-br committed:
MAINT: fix up several doctests

ev-br committed:
MAINT: fix minor hiccups in tutorials, add doc/user/conftest.py

The conftest is needed to get the smoke-docs configuration from
the main conftest.

To run doctesting, use (this is scipy's `smoke-tutorial`)

$ pytest --doctest-glob=*rst ~/repos/numpy/doc/source/user/ -v

ev-br committed:
appease the linter

ev-br committed:
ENH: add spin smoke-docs command

ev-br commented:
Okay, the last commit adds a `spin smoke-docs` command --- the name follows the scipy's `dev.py smoke-docs`. The user API is exactly the same as `spin test`, only it only runs doctests:

```
$ spin smoke-docs      # == python runtests.py refguide-check
$ spin smoke-docs numpy/lib
$ spin smoke-docs -t numpy/linalg/_linalg.py -v
# ....
```

Curiously, somewhere in the spin machinery warnings get turned into errors by default. I wonder where this switch is.
Other than that, the direct `pytest  --doctest-modules` invocation passes locally for me.

ev-br committed:
MAINT: smoke-docs: filter out all deprecation and runtime warnings

ev-br committed:
CI: run spin smoke-docs on CircleCI

ev-br committed:
appease the linter

ev-br committed:
CI: move smoke-docs run to GHA

ev-br committed:
CI: smoke-docs: install test/doctest deps (MPL, hypothesis)

ngoldbaum commented:
Thanks so much for working on this.

I know at some point @seberg was looking at adding docstring checking with pytest-doctest-plus. Just a head's up in case that intersects with this change.

Can we get a mention of this in the developer guide too?

ev-br commented:
Thanks Nathan!
It does intersect indeed :-).
The demo is here; from my side what's left are a few minor tweaks to 
- properly run it on CI;
- port smoke-testing the *rst tutorials (scipy's `dev.py smoke-tutorials`);
- rip out the doctesting part of tools/refguide-check.py;
- update the devguide

I think at this stage I'm going to pause working on this, and ask for an exec decision.

seberg commented:
> I know at some point @seberg was looking at adding docstring checking with pytest-doctest-plus. Just a head's up in case that intersects with this change.

Yeah, it would be nice to consolidate the ecosystem more and doctestplus seemed/seems like the more likely candiate for that, but I don't think that should stop improvements here either way.

ev-br committed:
MAINT: clean up `spin smoke-docs`

Follow what pytest --doctest-modules allows:
- directories
- -k patterns

`pytest --doctest-modules path/to/py/file` I'd expect to work,
but apparently it fails to collect anything --- and this is
not related to scipy-doctests.

ev-br committed:
DOC: random: do not skip import scipy if it is later used

ev-br committed:
CI: install doctest dependencies

ev-br committed:
TST: doctests: drop dtype(...) into the check namespace

ev-br committed:
TST: smoke-docs: skip __array_namespace_info__

ev-br commented:
Okay, the PR is ready from my side.

- CI is hooked up to the `benchmark` run to save a numpy build; can move it to a more appropriate place if wanted;
- some skips can be fixed in the source; I only fixed the "obvious" ones, the rest are an enhancement (anything that worked under refguide-check keeps working, so any failures are "new")
- tutorials can be done in a follow-up
- `spin` integration is

```
$ spin smoke-docs
$ spin smoke-docs numpy/linalg
$ spin smoke-docs -- -k"det and not slogdet"
```

asmeurer commented:
As I noted in another issue, I would suggest using a more standard name for this to improve the usability, like `spin doctest` or just `spin test --doctest` (and if the latter, I would also consider making it run by default).

seberg commented on numpy/conftest.py:
```diff
@@ -136,3 +144,72 @@ def weak_promotion(request):
     numpy._set_promotion_state(state)
+
+
+if HAVE_SCPDT:
```
I don't really think (not quite sure) that this whole pattern is the pytest way to do things.

`warnings_errors_and_rng` looks like it could be a `hook` on the plugin side, most other configs can probably be part of `pyproject.toml`, although I suspect one can set them also in the `conftest.py` (even if `scipy-doctest` isn't available).

ev-br replied:
The context manager technically could be a pytest hook. I would not do it though, since it's breaks the separation between the doctest and pytest layers. 
Setting them unconditionally is a tad strange: these are specific for `scipy-doctest`.

seberg reviewed (commented):
One thing is that I am not sure whether our docs pass on all (64bit) systems or if e.g. on windows some results will just be different.  But maybe we shouldn't worry about it.

The CircleCI runner is using other infrastructure to run the doctests now, is there a difference between the two?!

Overall looks fine to me and I am happy with putting it in (with some tweaks).  Running with `pytest` is an improvement.

However, I will note that unless there are some serious problems with it aligning with pytest-doctestplus seems like the better long term solution.
(Yes, adopting it is harder/churn due it behaving more sensitive or just a bit different.  But it is used by quite a few projects, and unlike us, many of those projects actually rely on it for testing their code;  I don't condone that, but I suspect it gives them more need to maintain it well)

ev-br commented:
The name of the spin command follows an earlier discussion of the scipy version: https://github.com/scipy/scipy/pull/20127#issuecomment-1957062024 (@tupui's suggestion)
I still think that mixing doctests with unit tests is a bad idea.

Re: this or pytest-doctestplus: I'm fine either way, just would really prefer to not spend an effort crossing the t-s and dotting the i-s if in the end it's not wanted.

mattip commented:
> I still think that mixing doctests with unit tests is a bad idea.

Is there another discussion of this position somewhere? If not could you elaborate? In my mind, doctests are just another form of unit test, and I often peruse the unit tests as extended examples beyond the documentation

ev-br committed:
MAINT: smoke-docs: move ignores to conftest

ev-br commented on numpy/conftest.py:
```diff
@@ -136,3 +144,72 @@ def weak_promotion(request):
+    dt_config.pytest_extra_xfail = {
+        'how-to-verify-bug.rst': '',
+        'c-info.ufunc-tutorial.rst': '',
+        'basics.interoperability.rst': 'needs pandas',
```
I don't believe there is a conceptual difference. Both pytest-doctestplus and scipy-doctest do check rst. 
The pytest incantation may be slightly different (for scipy-doctest it's `$ pytest --doctest-glob=*rst ~/repos/numpy/doc/source/user/ -v`, just not packaged to spin yet and I'm not sure it needs to be) but that's not a big deal either way.

What is common though is that e.g. `how-to-verify-bug.rst` is not meant for doctesting as written. So whichever tool doctests it, it'll either need  a bunch of skips in the text, or a global skip in the tool.

And I believe that if the docs are for the human reader, skips in the text is visual clutter at best.

ev-br commented on numpy/exceptions.py:
```diff
@@ -222,7 +222,9 @@ class DTypePromotionError(TypeError):
     promoted:
 
-    >>> np.result_type(np.dtype("M8[s]"), np.complex128)
+    >>> np.result_type(np.dtype("M8[s]"), np.complex128)  # doctest: +IGNORE_EXCEPTION_DETAIL
```
Yes it's needed. 
Note that the scipy version has `+IGNORE_EXCEPTION_DETAIL` by default, but numpy removed it at some point.
I'd advocate for adding it by default, but this PR tries to be minimal :-).
Without it, 

```
225     >>> np.result_type(np.dtype("M8[s]"), np.complex128)
Differences (unified diff with -expected +actual):
    @@ -1,7 +1,6 @@
     Traceback (most recent call last):
    - ...
    -DTypePromotionError: The DType <class 'numpy.dtype[datetime64]'> could not
    -be promoted by <class 'numpy.dtype[complex128]'>. This means that no common
    -DType exists for the given inputs. For example they cannot be stored in a
    -single array unless the dtype is `object`. The full list of DTypes is:
    -(<class 'numpy.dtype[datetime64]'>, <class 'numpy.dtype[complex128]'>)
    +  File "/home/br/mambaforge/envs/numpy-dev/lib/python3.11/doctest.py", line 1355, in __run
    +    exec(compile(example.source, filename, "single",
    +  File "<doctest numpy.exceptions.DTypePromotionError[0]>", line 1, in <module>
    +    np.result_type(np.dtype("M8[s]"), np.complex128)
    +numpy.exceptions.DTypePromotionError: The DType <class 'numpy.dtypes.DateTime64DType'> could not be promoted by <class 'numpy.dtypes.Complex128DType'>. This means that no common DType exists for the given inputs. For example they cannot be stored in a single array unless the dtype is `object`. The full list of DTypes is: (<class 'numpy.dtypes.DateTime64DType'>, <class 'numpy.dtypes.Complex128DType'>)
```

seberg commented:
Well, we never used doctests to check the *code* for correctness only to check the *docs* for correctness.  That is right IMO since docs should change freely without worrying about code test coverage.

OTOH, that argument doesn't mean you can't just run the doctests always: docs getting out-dated isn't much better than code tests starting to fail.

However, I suspect doctests are more flaky (e.g. we never run them on 32bit platforms, I bet and I am not sure if they might be python version sensitive more than other tests).
So, I am fine either way, but someone might have to be a bit dedicated to ensure it's not flaky?

> The name [smoke-docs] of the spin command follows earlier discussion of the scipy version

Thanks, but I suppose I don't see that `spin smoke-docs` is better than `spin doctest`.

---

I am OK with this largely as it is, it isn't like the `spin` invocation would change if we would put in doctestplus.

ev-br commented:
> perhaps we should make a decision which way we wish to go. #21070 seems to have a few parallel discussions:
>
> - how to run doctests
> - how to get feedback which doctests have run
> - whether to pivot from refguide-check to doctest-plus or some other scientific-python wide solution

To summarize main high-level difference between scipy-doctest and doctestplus (+opinionated comments):

Architecture: 
- scipy-doctest is more modular: it has a clear separation between doctest layer and the pytest plugin;
- the core doctesting layer is also modular: can easily mix-and-match the Checker, Finder and Parser. For one, can reuse the doctestplus Checker if wanted (https://github.com/scipy/scipy_doctest/blob/main/scipy_doctest/tests/test_runner.py#L98). 
- the core Checker tech is somewhat different: one is `np.allclose`, the other is regex (I believe this is minor, both techs are well established)

UI:

- human-readable markers, `# may vary`, are superior to `# doctest: +FLOAT_CMP`, IMO
- Flexibility in doctest finding (e.g. only doctest public objects) is handy and has long history in both numpy and scipy
- Flexibility in skipping things is essential  (consider `plt.xlim([1, 2])  # doctest: +SKIP`)
- Having to tweak whitespace to be exactly right is annoying (doctests's `NORMALIZE_WHITESPACE` is not sufficient).

Overall, all features of scipy-doctest can be added to doctestplus. I'm on record suggesting an ecosystem wide solution, https://github.com/numpy/numpy/issues/21070#issuecomment-1133832331. That discussion went nowhere though, which is why there is now a `scipy-doctest` plugin :-).

ev-br commented:
>> I still think that mixing doctests with unit tests is a bad idea.

> Is there another discussion of this position somewhere? If not could you elaborate? In my mind, doctests are just another form of unit test, and I often peruse the unit tests as extended examples beyond the documentation

There is some discussion in https://github.com/scipy/scipy/pull/20127.

Perusing unit tests as extended examples --- definitely yes, it's likely not just you and me doing this regularly :-).
Going the other way --- using doctests to test functionality instead of unit tests --- is what I think is a bad idea. The reason is basically that documentation is for a reader, so the main goal of docstring examples are to demonstrate things to a reader. Doctesting them is just a technical implement to keep examples from bitrotting, no more.
IOW, docstring examples are for the users; unit tests-as-documentation is for developers.

ev-br commented on .spin/cmds.py:
```diff
@@ -258,6 +258,86 @@ def test(ctx, pytest_args, markexpr, n_jobs, tests, verbose, *args, **kwargs):
+        '--ignore=numpy/_pyinstaller',
+        '--ignore=numpy/random/_examples',
+        '--ignore=numpy/compat',
+        '--ignore=numpy/f2py/_backends/_distutils.py',
```
Done in cd69c4dc35dc3c4ad5cccc1eebc95ea9ba1088cd.

mattip commented:
I think `scipy-doctest` should be added to `requirements/test_requirements.txt`, otherwise running this locally one gets a pytest error about a missing `--doctest-collect` command line argument

ev-br committed:
MAINT: add scipy-doctest to requirements/test_requirements.txt

ev-br commented:
Added to `test_requirements.txt`  (before seeing the edit). test_requirements make sense to me more than build_requirements.txt (next to pytest-xdist, also a pytest plugin?), but am happy to move if wanted.

ev-br committed:
rerun the CI

mattip committed:
change smoke-docs -> check-docs

mattip reviewed (approved):
We discussed this at the triage meeting, and this got a thumbs-up. The only nit is if we could change the name to check-docs. I took the liberty of changing the name.

mattip commented:
Just to be clear: this does not completely replace refguide-check since it does not check the rst documentation, correct? So the [circle ci job running refguide-check](https://github.com/numpy/numpy/blob/38d3d76c7c57186d979592c67a1e6e6239f0cbd2/.circleci/config.yml#L104) should still be running as previously.

ev-br commented:
Indeed. This is on my to do list. Scipy's ui is `dev.py smoke-tutorials`, so you want `spin check-tutorials` here then?

ngoldbaum commented:
> so you want spin check-tutorials here then?

Sounds fine to me, good to keep it consistent.

> also, this PR or a follow-up?

A followup is fine if you're ok with that. I'd like to be able to have the new `spin` subcommand ASAP! Every once in a while I need to figure out how to run the doctests and it's not very fun.

asmeurer commented:
I agree with Sebastian that doctests are not tests. They are examples that happen to be tested. A good documentation example is not a good test and a good test is not a good documentation example. They shouldn't be confused with each other. The point of doctests is just to ensure documentation examples actually work the way they say they do.

But that seems irrelevant to the question of whether something like `spin test` should automatically run the doctests. In my view, that's the simple command that contributors will run on their machine to ensure all their changes are working before they commit. It shouldn't require ten different commands with obscure options to do this (c.f. #26588).

ev-br commented:
So how about landing it as is, and I'll work on a follow-up to
- add `check-turorials`
- remove doctesting from the refguide-check.py script

Not sure how to make `spin test` run multiple commands. Surely it's doable.

seberg commented:
> But that seems irrelevant to the question of whether something like `spin test` should automatically run the doctests

I honestly don't care for that and if you want something that checks everything, it may have to also include linting?  And that would make it a new invocation, which is probably fine if we can think of it, like `comprehensive-check` ;).
Also, I don't trust that doc tests are not brittle, bunching it up might just make someone run them on niche platforms where it is a lost cause.

> it does not check the rst documentation, correct?

Ah, I somewhat expected it would.  FWIW, I might be happy to include the `rst` docs into `check-docs`, but either way works and maybe it's slow.

mattip commented:
> So how about landing it as is, and I'll work on a follow-up to
> -   add check-tutorials
> -   remove doctesting from the refguide-check.py script

Sounds like a plan, thanks!

mattip commented:
Thanks @ev-br

mattip commented:
FWIW: I view doctest as a way to test the documentation for correctness. So it is a "test". Documentation can have bugs just like code can have bugs, and testing is just one tool to expose those bugs. Relegating documentation tests to a category that is "not testing" does a disservice to the hard work put into making the documentation.

That said, it is true the examples in the documentation are meant to be "happy path" demonstrations of correct use of the functionality, and should not be taken as comprehensive tests to prove correctness of the code, so the goal of examples is not the same as the goal of unit tests.

asmeurer commented:
> I view doctest as a way to test the documentation for correctness. 

Yes, I agree completely, and that's part of my argument that this should be easier to invoke. I just want to push back from the idea that a doctest is *a test*, because that leads to people writing bad documentation examples (not to accuse anyone here of saying that, but I've seen it happen before).

dhomeier commented on numpy/_core/_add_newdocs.py:
```diff
@@ -6937,19 +6937,19 @@ def refer_to_array_attribute(attr, method=True):
 
     >>> np.array([1.2, object(), "hello world"],
-                 dtype=StringDType(coerce=True))
+    ...          dtype=StringDType(coerce=True))
```
This is taken over from the initial version, but should this not be `coerce=False`? – see also https://github.com/numpy/numpy/blob/a1f2d582f84878b6c67bd641fa671a4cf868fe8c/doc/source/user/basics.strings.rst?plain=1#L185-L188

seberg replied:
Yes, but it seems already fixed.

dhomeier replied:
- Sorry, must have looked at the wrong branch then – in `main` and 2.2.x it's still `True`.

seberg replied:
Sorry, confusing that this is twice in the docs, and only once is wrong.  Yeah, should just fix it...

dhomeier replied:
I was a bit puzzled that it did not raise in doctests, but apparently the exception was quoted incompletely.

ngoldbaum replied:
As the likely author of the typo (didn't check) thanks for noticing and fixing 😀
